File: hyemin_1/Permutation.java
class Permutation {
	void perm(int list[], int startIndex, int endIndex){
		int i, tmp;
		Swap swp = new Swap();
		
		/* startIndex가 endIndex와 같으면(즉, 마지막 인덱스까지 perm이 진행되면)
		 * list 숫자들 출력*/
		if(startIndex==endIndex){
			for(i=0; i<=endIndex; i++){
				System.out.print(list[i]+" ");
			}
			System.out.println();
		}
		
		//마지막 인덱스에 도달하지 않은 상태일 경우
		else{
			for(i=startIndex; i<=endIndex; i++){
				swp.swap(list, startIndex, i);//startIndex와 i번째 인덱스 swap
				perm(list, startIndex+1, endIndex);//startIndex 값 고정한 채로 perm 호출 
				swp.swap(list, startIndex, i);//list 값 원상복구
			}
		}
	}
}

File: permutation.cpp
#include <iostream>
using namespace std;

class Permutation {
public:
	void perm (char * list, int i, int n);
};

int main () {
	Permutation p;
	char list[] = {'a', 'b', 'c', 'd'};
	p.perm(list, 0, 3);
}

void Permutation::perm(char *list, int i, int n) {
	int j, temp;
	if (i==n) {
		for (j=0; j<=n; j++) {
			printf("%c", list[j]);
			printf("   ");
		}
		printf("\n");
	}
	else {
		for (j=i; j<=n; j++) {
			temp = list[i];
			list[i]=list[j];
			list[j]=temp;
			perm(list, i+1, n);
			temp = list[i];
			list[i]=list[j];
			list[j]=temp;
		}
	}
}